### PR TITLE
Bug fixes

### PR DESCRIPTION
**Pull request by cfognom**
- Fixed a bug that made commands not available in some contexts. For example in .xml files.
- Fixed a bug that caused DuplicateLinesUp/Down to insert an additional newline when it was not needed.

### CODE REVIEW

**Comment by justcla** [comment]
Thanks for the fixes. I notice that you changed the algorithm to calculate the last line by checking if "Environment.NewLine", but I'm concerned that might not cover all scenarios of file endings - ie. CR only, CRLR, LR only (ie. Windows endings vs Unix endings)
Have you tested this against all type of file endings?

I didn't notice the bug on the existing implementation. Can you describe a repro I can do on v4.3.2 so I can see what bug it fixes?

Thanks

**Comment by justcla** [comment]
Thanks for trying to enable the commands in XML. My concern is that "| SUPPORTED" will mean the commands show up on the menu in full colour - instead of greyed out - when the command is disabled. It's not a terrible scenario. Worst case is the command will look like it works, but it won't do anything. But, I'm not sure if we want to enable commands if they are only supported, but not enabled. But I'm happy to hear your thoughts on that. Thanks

**Comment by cfognom** [comment]
> Thanks for the fixes. I notice that you changed the algorithm to calculate the last line by checking if "Environment.NewLine", but I'm concerned that might not cover all scenarios of file endings - ie. CR only, CRLR, LR only (ie. Windows endings vs Unix endings) Have you tested this against all type of file endings?
> 
> I didn't notice the bug on the existing implementation. Can you describe a repro I can do on v4.3.2 so I can see what bug it fixes?
> 
> Thanks

You are right, I should have explained the bug more clearly. You can repro it by running duplicate lines up/down on the second last line in the document, while the last line is empty. See screenshots below.
Before command:
223491b0e31348af74b2f69960806eea
After command:
5ca7fdc780618ba60ff10309f7d94493
It has an additional newline in between.

Your concerns about "Environment.NewLine" are valid, I'll see if I can think of another way to fix it. I have not tested it for all types of line endings.

**Comment by cfognom** [comment]
> Thanks for trying to enable the commands in XML. My concern is that "| SUPPORTED" will mean the commands show up on the menu in full colour - instead of greyed out - when the command is disabled. It's not a terrible scenario. Worst case is the command will look like it works, but it won't do anything. But, I'm not sure if we want to enable commands if they are only supported, but not enabled. But I'm happy to hear your thoughts on that. Thanks

I was trying to figure out why the DuplicateLinesUp/Down commands was not working in XML documents. I saw 'QueryStatus' being triggered for the HotCommands Guid but not the 'Exec' method. Adding 'OLECMDF.OLECMDF_SUPPORTED' seemed to make it work. I don't 100% understand why C# contexts only require 'OLECMDF.OLECMDF_ENABLED' but XML contexts also need 'OLECMDF.OLECMDF_SUPPORTED'. All I know is that all examples I saw of IOleCommandTarget always used both to enable their commands.

**Comment by cfognom** [comment]
> I tested the Enabled/Supported behaviour on an XML. Yes, it enables the duplicate commands nicely (they are text based), but it also enabled the Move Member and other commands that fail with errors if it's not a C# file. Could you please adjust the logic so that most commands work only when Enabled, and let the Duplicate commands (and any others that work on text) come alive when Supported or Enabled? Thanks

That seems reasonable

**Comment by cfognom** [comment]
After researching a bit more about the QueryStatus method. I find it very strange how it could work without the 'OLECMDF.OLECMDF_SUPPORTED' flag set. Since this flag is used to indicate that the command is supported by the current context. Not setting it should mean the command can not work.

The 'OLECMDF.OLECMDF_ENABLED' flag on the other hand is used to indicate is the command is enabled. It is possible that a command is supported by the context but disabled for some reason and then it should not work either.

Am I misunderstanding their meanings?

**Comment by justcla** [comment]
Some commands are supported. And some supported commands are enabled.
As you said, a command might be supported, but not enabled.

Supported is redundant when checking if enabled.

But you added "Or supported". So that lights up all the commands, whether they are enabled or not.

**Comment by cfognom** [comment]
> Some commands are supported. And some supported commands are enabled. As you said, a command might be supported, but not enabled.
> 
> Supported is redundant when checking if enabled.
> 
> But you added "Or supported". So that lights up all the commands, whether they are enabled or not.

But the 'QueryStatus' method is what is determining the command status. They are not enabled from some other place, right? By ORing the flags together we are saying: "The command is enabled AND supported". By using only enabled we are saying: "The command is enabled but not supported" (That's why I find it weird how it could work when only setting the enabled flag) What seems to be recommended is to always set both flags together.

Perhaps in this case it would be better to determine if we are in a C# context and then not set the enabled + supported flags for the C# commands. Basically we either set both at the same time or none.

**Comment by justcla** [comment]
I see what you're saying. My bad. I'll take a better look.

**Comment by cfognom** [comment]
I made 2 more commits. They should address your concerns.

**Comment by cfognom** [comment]
Only potential problem I can see is that we now require C# content type for the commands whereas before it was undefined. It could mean that for certain content types the commands will be disabled where they previously were not. Do you think it was possible that some commands were working in non-C# contexts and users used them?

**Comment by justcla** [comment]
> I was trying to figure out why the DuplicateLinesUp/Down commands was not working in XML documents.

One thing I can mention is that the XMLEditor does not fall in hierarchy under TextEditor for shortcut bindings.
There is a hierarchy for commands: Global -> TextEditor -> C# Editor / C++ / F#, etc.
So if you bind a shortcut to TextEditor, it applies to all TextEditors plus C3, F3, C++,etc unless otherwise overridden at that level.

But that should not affect the command appearing on the menu under Edit->Advanced.

But - the XMLEditor is different. I don't know what's going on under the hood.

**Comment by justcla** [comment]
> Do you think it was possible that some commands were working in non-C# contexts and users used them?

Yes, I think it's very possible.
For example, in the vssettings file, I defined Edit.GoToLastEditLocation at the Global scope.
The other commands have shortcuts setup for "TextEditor" (which does not apply to XML!).

**Comment by justcla** [comment]
Thanks for the updates.

ToggleCommentCmdId should work anywhere "Comment" works.
ExpandSelectionCmdId should work on anything text-based.
FormatCodeCmdId should  work on anything with some language context.

Would be nice to have all these working on plain text (and XML)

MoveMember/GoToMember should work on all managed languages: C#, C/C++, F#.
But, I have never personally tested it on any of those. I don't mind if we limit it to C#. But maybe we can get it working on all managed languages.

**Comment by justcla** [comment]
Visibility Contraints defined in the VSCT file are currently limiting all these commands to TextEditor.
Given XMLEditor is not in that hierarchy, that might be causing a problem.

`  <VisibilityConstraints>
    <VisibilityItem guid="guidHotCommandsPackageCmdSet" id="cmdidToggleComment" context="GUID_TextEditorFactory"/>
    <VisibilityItem guid="guidHotCommandsPackageCmdSet" id="cmdidExpandSelection" context="GUID_TextEditorFactory"/>
    <VisibilityItem guid="guidHotCommandsPackageCmdSet" id="cmdidShrinkSelection" context="GUID_TextEditorFactory"/>
    <VisibilityItem guid="guidHotCommandsPackageCmdSet" id="cmdidFormatCode" context="GUID_TextEditorFactory"/>
    <VisibilityItem guid="guidHotCommandsPackageCmdSet" id="cmdidDuplicateLinesUp" context="GUID_TextEditorFactory"/>
    <VisibilityItem guid="guidHotCommandsPackageCmdSet" id="cmdidDuplicateLinesDown" context="GUID_TextEditorFactory"/>
    <VisibilityItem guid="guidHotCommandsPackageCmdSet" id="cmdidDuplicateSelection" context="GUID_TextEditorFactory"/>
    <VisibilityItem guid="guidHotCommandsPackageCmdSet" id="cmdidDuplicateSelectionReverse" context="GUID_TextEditorFactory"/>
    <VisibilityItem guid="guidHotCommandsPackageCmdSet" id="cmdidJoinLines" context="GUID_TextEditorFactory"/>
    <VisibilityItem guid="guidHotCommandsPackageCmdSet" id="cmdidMoveMemberUp" context="GUID_TextEditorFactory"/>
    <VisibilityItem guid="guidHotCommandsPackageCmdSet" id="cmdidMoveMemberDown" context="GUID_TextEditorFactory"/>
    <VisibilityItem guid="guidHotCommandsPackageCmdSet" id="cmdidGoToPreviousMember" context="GUID_TextEditorFactory"/>
    <VisibilityItem guid="guidHotCommandsPackageCmdSet" id="cmdidGoToNextMember" context="GUID_TextEditorFactory"/>
  </VisibilityConstraints>`

**Comment by cfognom** [comment]
ExpandSelectionCmdId crashes when ran on XML document, so best to restrict it to C#.
ToggleCommentCmdId seems to work for other content types, maybe we can enable it for 'Code' contents? (See commit [3f2fc7d](https://github.com/justcla/HotCommands/pull/100/commits/3f2fc7d46991464c6655875d6bb865585ed543c4))
I could not verify that FormatCodeCmdId works for non-C# contents, I'll keep it in C#.
I have not been able to test the member commands, so I'll keep those in C# too.

Resource to see available content types: https://github.com/MicrosoftDocs/visualstudio-docs/blob/main/docs/extensibility/language-service-and-editor-extension-points.md

**Comment by justcla** [comment]
Nice. Thanks for exploring each of the commands.

**Comment by cfognom** [comment]
> Visibility Contraints defined in the VSCT file are currently limiting all these commands to TextEditor. Given XMLEditor is not in that hierarchy, that might be causing a problem.
> 
> `
> 
> ```
> <VisibilityItem guid="guidHotCommandsPackageCmdSet" id="cmdidToggleComment" context="GUID_TextEditorFactory"/>
> <VisibilityItem guid="guidHotCommandsPackageCmdSet" id="cmdidExpandSelection" context="GUID_TextEditorFactory"/>
> <VisibilityItem guid="guidHotCommandsPackageCmdSet" id="cmdidShrinkSelection" context="GUID_TextEditorFactory"/>
> <VisibilityItem guid="guidHotCommandsPackageCmdSet" id="cmdidFormatCode" context="GUID_TextEditorFactory"/>
> <VisibilityItem guid="guidHotCommandsPackageCmdSet" id="cmdidDuplicateLinesUp" context="GUID_TextEditorFactory"/>
> <VisibilityItem guid="guidHotCommandsPackageCmdSet" id="cmdidDuplicateLinesDown" context="GUID_TextEditorFactory"/>
> <VisibilityItem guid="guidHotCommandsPackageCmdSet" id="cmdidDuplicateSelection" context="GUID_TextEditorFactory"/>
> <VisibilityItem guid="guidHotCommandsPackageCmdSet" id="cmdidDuplicateSelectionReverse" context="GUID_TextEditorFactory"/>
> <VisibilityItem guid="guidHotCommandsPackageCmdSet" id="cmdidJoinLines" context="GUID_TextEditorFactory"/>
> <VisibilityItem guid="guidHotCommandsPackageCmdSet" id="cmdidMoveMemberUp" context="GUID_TextEditorFactory"/>
> <VisibilityItem guid="guidHotCommandsPackageCmdSet" id="cmdidMoveMemberDown" context="GUID_TextEditorFactory"/>
> <VisibilityItem guid="guidHotCommandsPackageCmdSet" id="cmdidGoToPreviousMember" context="GUID_TextEditorFactory"/>
> <VisibilityItem guid="guidHotCommandsPackageCmdSet" id="cmdidGoToNextMember" context="GUID_TextEditorFactory"/>
> ```
> 
> `

Even though there are these constraints the commands are still visible and can be used from XML documents with the recent changes. I'm not sure how these constraints affect the command visibility.

**Comment by justcla** [comment]
I've tested it. Looks pretty good. Nice job. Thanks!
Duplicate commands now work in all files. That's a neat improvement!

ToggleComment only seems to do "Comment" (never does Uncomment). But my implementation of that command was very weak and can be improved in a different issue (if ever).

I'll go ahead and merge and release this version. Thanks again.

**Comment by justcla** [comment]
It is released on the Marketplace now as v4.3.3

**Comment by cfognom** [comment]
I think I understand what the visibility constraints are doing! They are simply another source when determining if a command is available. So a command can be either enabled from 'QueryStatus' or from the VSCT file's visibility constraints. If it is enabled in one and disabled in the other it becomes enabled, logical OR.

I did a test: I removed all visibility constraints while using v4.3.2's 'QueryStatus' logic that only set enabled and not supported => no command is available. Essentially, the method did nothing in v4.3.2! All status information came from VSCT file. Only when adding the enabled flag did the method start marking commands as available.

This means that, since, as I understand it, 'Code' and 'CSharp' content types are a subset of the VSCT's context="GUID_TextEditorFactory", those commands will use enabled logic from VSCT like the always have. But 'any' content type is a superset of context="GUID_TextEditorFactory" so the duplicate commands will be enabled in a superset of situations.

In conclusion, the modifications to 'QueryStatus' did strictly increase the availability of Duplication commands while keeping the availability of all other commands.

**Comment by justcla** [comment]
Excellent. Thanks for the analysis and explanation!
